compiler/rustc_trait_selection/src/traits/util.rs

```diff
@@ -507,19 +509,72 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for PlaceholderReplacer<'_, 'tcx> {
 }
 
 pub fn sizedness_fast_path<'tcx>(tcx: TyCtxt<'tcx>, predicate: ty::Predicate<'tcx>) -> bool {
-    // Proving `Sized` very often on "obviously sized" types like `&T`, accounts for about 60%
-    // percentage of the predicates we have to prove. No need to canonicalize and all that for
-    // such cases.
+    // Proving `Sized`/`MetaSized`/`PointeeSized`, very often on "obviously sized" types like
+    // `&T`, accounts for about 60% percentage of the predicates we have to prove. No need to
+    // canonicalize and all that for such cases.
     if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_ref)) =
         predicate.kind().skip_binder()
     {
-        if tcx.is_lang_item(trait_ref.def_id(), LangItem::Sized)
-            && trait_ref.self_ty().is_trivially_sized(tcx)
-        {
+        let sizedness = if tcx.is_lang_item(trait_ref.def_id(), LangItem::Sized) {
+            SizedTraitKind::Sized
+        } else if tcx.is_lang_item(trait_ref.def_id(), LangItem::MetaSized) {
+            SizedTraitKind::MetaSized
+        } else if tcx.is_lang_item(trait_ref.def_id(), LangItem::PointeeSized) {
+            SizedTraitKind::PointeeSized
+        } else {
+            return false;
+        };
+
+        if trait_ref.self_ty().has_trivial_sizedness(tcx, sizedness) {
             debug!("fast path -- trivial sizedness");
             return true;
         }
     }
 
     false
 }
+
+/// To improve performance, sizedness traits are not elaborated and so special-casing is required
+/// in the trait solver to find a `Sized` candidate for a `MetaSized` obligation. Returns the
+/// predicate to used in the candidate for such a `obligation`, given `candidates`.
+pub(crate) fn unelaborated_sizedness_candidate<'tcx>(
```

Comment:
`fn prove_meta_sized_via_elaborated_sized`

pls only take a single candidate per call

Reply:
Fixed, didn't rename as it is now returning the `candidate`, but I can if we want to.